src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -1105,27 +1106,37 @@ sealed class RsDiagnostic(
             )
     }
 
-    class MissingFieldsInTuplePattern(pat: RsPat, private val declaration: RsFieldsOwner) : RsDiagnostic(pat) {
+    class MissingFieldsInTuplePattern(
+        pat: RsPat,
+        private val declaration: RsFieldsOwner,
+        private val expectedAmount: Int,
+        private val actualAmount: Int
+    ) : RsDiagnostic(pat) {
         override fun prepare(): PreparedAnnotation {
             val itemType = if (declaration is RsEnumVariant) "Enum variant" else "Tuple struct"
+            val fieldOrFields = if (expectedAmount == 1) "field" else "fields"
             return PreparedAnnotation(
                 ERROR,
                 E0023,
-                // TODO: provide number of expected/actual fields
-                "$itemType pattern does not correspond to its declaration",
+                "$itemType pattern does not correspond to its declaration: Expected $expectedAmount $fieldOrFields, found $actualAmount",
                 fixes = listOf(AddStructFieldsPatFix(element))
             )
         }
     }
 
-    class MissingFieldsInStructPattern(pat: RsPat, private val declaration: RsFieldsOwner) : RsDiagnostic(pat) {
+    class MissingFieldsInStructPattern(
+        pat: RsPat,
+        private val declaration: RsFieldsOwner,
+        private val missingFields: List<RsFieldDecl>
+    ) : RsDiagnostic(pat) {
         override fun prepare(): PreparedAnnotation {
             val itemType = if (declaration is RsEnumVariant) "Enum variant" else "Struct"
+            val missingFieldNames = missingFields.map { "`${it.name!!}`" }.stream().collect(Collectors.joining(", "))
```

Comment:
```suggestion
            val missingFieldNames = missingFields.joinToString(", ") { "`${it.name!!}`" }
```

Reply:
Also, could you check that `it.name!!` doesn't fail with incomplete code (the corresponding test will be helpful here as well)

Reply:
> Also, could you check that `it.name!!` doesn't fail with incomplete code (the corresponding test will be helpful here as well)?

I have added a test for error analysis on invalid syntax but I'm not sure if it is correct solution so could you recheck it, please?
https://github.com/intellij-rust/intellij-rust/pull/4315/files#diff-4c22a3080d6a06f7e6d2b6a38f801ab0R2899-R2905